### PR TITLE
DRAFT: Add msvc support

### PR DESCRIPTION
**Pull request by phcoder**
This is ugly. Do we want it or should we just skip msvc?

### CODE REVIEW

**Comment by phcoder** [comment]
@leiradel your opinion on this?

File: gwlua/bsreader.c [hunk 1]
```diff
@@ -61,15 +61,15 @@ static size_t getliteral( stream_t* stream, char* literal, size_t size )
 }
 
 const char* bsread( lua_State* L, void* data, size_t* size )
-{
-  (void)L;
```
Comment (leiradel) [inline gwlua/bsreader.c:65]: I'm curious what does moving this line fixes.

File: gwlua/functions.c [hunk 2]
```diff
@@ -355,13 +355,14 @@ static int l_setbackground( lua_State* L )
...
   
   int width = MAX( bg->width, 480 );
+  int x0;
   
   if ( rl_backgrnd_create( width, bg->height ) )
   {
     return luaL_error( L, "out of memory allocating the background framebuffer" );
   }
   
-  int x0 = ( width - bg->width ) / 2;
```
Comment (leiradel) [inline gwlua/functions.c:364]: Isn't it possible to tell MSVC to compile the source code as C99?

**Comment by leiradel** [comment]
This seems to have been merged in a rush. MSVC does seem to be able to build C99 code, which is almost 100% of the changes in this PR. The single other change introduces the possibility of a buffer overrun.

**Comment by twinaphex** [comment]
No, these changes are needed and necessary. MSVC2003/2005/2008/2010 need C89 support, they are not able to build C99 code at all.

File: gwlua/functions.c (hunk 2 above)
Comment (twinaphex) [reply gwlua/functions.c:364]: No, for these older versions it is not, and you should just declare variables at the top of a function or code block.

**Comment by leiradel** [comment]
@twinaphex [this page](http://users.ece.utexas.edu/~patt/20f.460n/labs/vstudio.html) suggests that the MS extensions makes the compiler understand C99, but I'm not the expert here.

**Comment by leiradel** [comment]
@twinaphex but the issue with using `strncpy` wasn't addressed, and still the code was merged.

**Comment by twinaphex** [comment]
> @twinaphex [this page](http://users.ece.utexas.edu/~patt/20f.460n/labs/vstudio.html) suggests that the MS extensions makes the compiler understand C99, but I'm not the expert here.

You're indeed not the expert here if you think that.

Anyway, if you're not going to be involved with this core anymore, I don't see why you need to interfere with our development process like this. We are free to make whatever decisions we want with the cores in the way we deem is best.

**Comment by leiradel** [comment]
No, I don't think I'm the expert here. Yes, you're free to open up the possibility of buffer overruns in the code.

**Comment by phcoder** [comment]
> The single other change introduces the possibility of a buffer overrun.
Can you elaborate? I think you refer to:
-      snprintf( name, sizeof( name ), "%s%s", button_name( i ), p == 0 ? "" : "/2" );
+      name[sizeof( name ) - 1] = '\0';
+      strncpy(name, button_name( i ), sizeof( name ) - 1);
+      if (p != 0)
+       strncat (name, "/2", sizeof( name ) - (strlen(name) + 1));
First call is limited to sizeof( name ) - 1 and second one adds at most sizeof( name ) - 1 - strlen(name). When can it overflow?

**Comment by leiradel** [comment]
I missed the `- 1` int the last argument for `strncpy`, so the code is good. Usually I use `strncpy` first, and then force a null terminator at the end of the string by setting it explicitly.

Sorry for the noise.

**Pull request merged**